ambry-network/src/main/java/com.github.ambry.network/Selector.java

```diff
@@ -347,13 +356,30 @@ public void poll(long timeoutMs, List<NetworkSend> sends)
           close(key);
         }
       }
+      completeSslHandshakes();
       this.metrics.selectorIORate.inc();
     }
     long endIo = time.milliseconds();
     this.metrics.selectorIOTime.update(endIo - endSelect);
   }
 
   /**
+   * Add those Ssl connections to connected list on handshake completion
+   */
+  private void completeSslHandshakes() {
+    Iterator<PendingHandshakeTransmission> pendingSslTransIter = pendingHandshakeTransmissions.values().iterator();
+    while (pendingSslTransIter.hasNext()) {
+      PendingHandshakeTransmission pendingSslTrans = pendingSslTransIter.next();
+      if (pendingSslTrans.sslTransmission.ready()) {
+        connected.add(pendingSslTrans.sslTransmission.getConnectionId());
+        metrics.selectorConnectionCreated.inc();
+        metrics.selectorPerceivedSslHandshakeTime.update(System.currentTimeMillis() - pendingSslTrans.pendingSinceMs);
+        pendingHandshakeTransmissions.remove(pendingSslTrans.sslTransmission.getConnectionId());
```

Comment:
Or maybe I am wrong. As per the docs, it seems the concurrent modification exception gets thrown even if the same threads modify the list structurally outside of the iterator, which kinda makes sense. Best to always use the iterator.